scripts/externalTests/test_helpers.py

```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------------------------------------------
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import os
+import re
+import subprocess
+import sys
+from abc import abstractmethod, ABCMeta
+from argparse import ArgumentParser
+from dataclasses import dataclass, field
+from enum import Enum
+from pathlib import Path
+from shutil import rmtree
+from tempfile import mkdtemp
+from textwrap import dedent
+from typing import List, Set
+
+# Our scripts/ is not a proper Python package so we need to modify PYTHONPATH to import from it
+# pragma pylint: disable=import-error,wrong-import-position
+PROJECT_ROOT = Path(__file__).parents[2]
+sys.path.insert(0, f"{PROJECT_ROOT}/scripts/common")
+
+from git_helpers import git_commit_hash
+
+SOLC_FULL_VERSION_REGEX = re.compile(r"^[a-zA-Z: ]*(.*)$")
+SOLC_SHORT_VERSION_REGEX = re.compile(r"^([0-9.]+).*\+|\-$")
+
+evm_version = os.environ.get("DEFAULT_EVM")
+CURRENT_EVM_VERSION: str = evm_version if evm_version is not None else "shanghai"
```

Comment:
```suggestion
CURRENT_EVM_VERSION: str = os.environ.get("DEFAULT_EVM", "shanghai")
```

Still, this mechanism looks unused. I see that `soltest_all.sh` uses such a variable but that script does not run external tests.

Doesn't hard-coding `shanghai` here defeat the purpose of using a variable?

Also, why not a command-line option?

Reply:
> Doesn't hard-coding `shanghai` here defeat the purpose of using a variable?

It does, but we need to have a default or maybe fail if none is provided.
 
> Also, why not a command-line option?

Yeah, I guess a command-line option makes more sense and we can just throw an error if it is not given.

Reply:
Ah, but note that we are using positional arguments here for compatibility reasons with the other external tests scripts (see: https://github.com/ethereum/solidity/blob/develop/test/external_tests.py#L52), so the `evm_version` and the `selected_presets` (you suggested here: https://github.com/ethereum/solidity/pull/13873#discussion_r1269390137) would also be a positional arguments.

Reply:
Actually, the `evm_version` is part of the config of the test, so it does not need to be a command-line option, we can just always require it to be set in the config. E.g.:
```python
test_config = TestConfig(
    name="PRBMath",
    repo_url="https://github.com/PaulRBerg/prb-math.git",
    ref_type="branch",
    ref="main",
    evm_version="shanghai",
    compile_only_presets=[],
    ...
)
```
And in each external test script we could read the current EVM version from the environment variables, like it is done by `soltest_all.sh`.

The disadvantage of this, compared with the current approach, is that you would need to potentially change the default value for the EVM version in many places instead of only change in `test_helpers.py`.

Reply:
After looking again in the code, I would just keep it like it is for now. I don't see much benefit to change it to a command-line option. It would require more refactors in the way we currently run the external tests, and I don't think it is worth do that now.

You can still setting an EVM version by setting the `DEFAULT_EVM` environment variable or passing `evm_version` in the external test config.

Reply:
I think I'd just remove the variable and use a hard-coded value then.

It would be better to have this value in CI but then it's harder to use the script manually (because you always have to specify EVM version).

Alternatively, we could just drop this and rely on compiler's default value. I'm not sure why we were not doing that in the original scripts.

Reply:
> Alternatively, we could just drop this and rely on compiler's default value. I'm not sure why we were not doing that in the original scripts.

Right. But how would we get the compiler's default value?

Reply:
We don't have to. If we just don't specify EVM version, it will use the default.

Reply:
Ah! I see. But then, how would we set in the foundry config for example?

Because the EVM version information is used to build the foundry profiles for prb-math.
For example:
```
[profile.legacy_optimize_evm_yul.optimizer_details]
yul = true
[profile.legacy_optimize_evm_only]
gas_reports = ["*"]
auto_detect_solc = false
solc = "/home/roqs/src/EF/solc-v0.8.20"
evm_version = "shanghai"
optimizer = true
via_ir = false
```
This is required because the default EVM version in the `prb-math` is `paris`.

So we need to have a way to inform the EVM version to the external test scripts.

Reply:
oh, I guess this might have been the reason why we're doing it. I.e. frameworks defaulting to older EVM.

In that case you'r right. We need to have the version hard-coded. I'd drop the variable though. If we're hardcoding the value anyway, it's more trouble than it's worth.

I mean, if you want to keep it I'd still prefer a command-line flag. We'll not be using that flag in CI so it not being compatible with the old script should not matter for that use case.

Reply:
Ok. By drop the variable you mean the `CURRENT_EVM_VERSION` right? And just hardcode `shanghai` in the `TestConfig`.

Reply:
Like, it is not that I don't like the command-line, but it would require some more refactors as far as I could see, but I may be wrong. But I think we could let to improve this in a later PR.

Reply:
No, I meant `DEFAULT_EVM`. `CURRENT_EVM_VERSION` is fine.

Reply:
> Like, it is not that I don't like the idea of making it a command-line option, but it would require some more refactors as far as I could see, but I may be wrong. But I think we could let this be improved in a later PR.

I think we can ignore the flag for now. I only suggested it in case you really wanted to keep the variable - i.e. I'd prefer flag over a variable.